File: python/test/t_FisherSnedecorFactory_std.py
#!/usr/bin/env python

from __future__ import print_function
import openturns as ot
import openturns.testing as ott
from math import sqrt

ot.PlatformInfo.SetNumericalPrecision(5)
distribution = ot.FisherSnedecor(4.5, 8.4)
size = 10000
sample = distribution.getSample(size)
factory = ot.FisherSnedecorFactory()
print('Distribution                      =', repr(distribution))
result = factory.buildEstimator(sample)
estimatedDistribution = result.getDistribution()
print('Estimated distribution            =', repr(estimatedDistribution))
parameterDistribution = result.getParameterDistribution()
print('Parameter distribution            =', parameterDistribution)
defaultDistribution = factory.build()
print('Default distribution              =', defaultDistribution)
fromParameterDistribution = factory.build(distribution.getParameter())
print('Distribution from parameters      =', fromParameterDistribution)
typedEstimatedDistribution = factory.buildAsFisherSnedecor(sample)
print('Typed estimated distribution      =', typedEstimatedDistribution)
defaultTypedDistribution = factory.buildAsFisherSnedecor()
print('Default typed distribution        =', defaultTypedDistribution)
typedFromParameterDistribution = factory.buildAsFisherSnedecor(
    distribution.getParameter())
print('Typed distribution from parameters=', typedFromParameterDistribution)
# Various estimators
estimatedDistribution = factory.build(sample)
print('Estimated distribution with default estimator =', estimatedDistribution)
estimatedDistribution = factory.buildMethodOfMoments(sample)
print('Estimated distribution with moments=', estimatedDistribution)
estimatedDistribution = factory.buildMethodOfLikelihoodMaximization(sample)
print('Estimated distribution with likelihoodMax.=', estimatedDistribution)

# Build method of moments
print("Build method of moments")
size = 10000
distribution = ot.FisherSnedecor(4.5, 8.4)
sample = distribution.getSample(size)
factory = ot.FisherSnedecorFactory()
estimatedTN = factory.buildMethodOfMoments(sample)
exact_mu = sample.computeMean()[0]
exact_sigma2 = sample.computeCovariance()[0,0]
computed_mu = estimatedTN.getMean()[0]
computed_sigma2 = estimatedTN.getCovariance()[0,0]
ott.assert_almost_equal(exact_mu, computed_mu, 1.e-15, 0.0)
ott.assert_almost_equal(exact_sigma2, computed_sigma2, 1.e-15, 0.0)

# Build method of likelihood maximization
print("Build method of likelihood maximization")
size = 10000
distribution = ot.FisherSnedecor(4.5, 8.4)
sample = distribution.getSample(size)
factory = ot.FisherSnedecorFactory()
estimatedTN = factory.buildMethodOfLikelihoodMaximization(sample)
exact_d1 = distribution.getD1()
exact_d2= distribution.getD2()
computed_d1 = estimatedTN.getD1()
computed_d2 = estimatedTN.getD2()
ott.assert_almost_equal(exact_d1, computed_d1, 0.0, 100.0 / sqrt(size))
ott.assert_almost_equal(exact_d2, computed_d2, 0.0, 100.0 / sqrt(size))

File: lib/test/t_FisherSnedecorFactory_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class FisherSnedecor for standard methods
 *
 *  Copyright 2005-2023 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    PlatformInfo::SetNumericalPrecision(5);
    FisherSnedecor distribution(4.5, 8.4);
    UnsignedInteger size = 10000;
    Sample sample(distribution.getSample(size));
    FisherSnedecorFactory factory;
    CovarianceMatrix covariance;
    // Distribution estimatedDistribution(factory.build(sample, covariance));
    Distribution estimatedDistribution(factory.build(sample));
    fullprint << "Distribution          =" << distribution << std::endl;
    fullprint << "Estimated distribution=" << estimatedDistribution << std::endl;
    // fullprint << "Covariance=" << covariance << std::endl;
    estimatedDistribution = factory.build();
    fullprint << "Default distribution=" << estimatedDistribution << std::endl;
    estimatedDistribution = factory.build(distribution.getParameter());
    fullprint << "Distribution from parameters=" << estimatedDistribution << std::endl;
    FisherSnedecor estimatedFisherSnedecor(factory.buildAsFisherSnedecor(sample));
    fullprint << "FisherSnedecor          =" << distribution << std::endl;
    fullprint << "Estimated fisherSnedecor=" << estimatedFisherSnedecor << std::endl;
    estimatedFisherSnedecor = factory.buildAsFisherSnedecor();
    fullprint << "Default fisherSnedecor=" << estimatedFisherSnedecor << std::endl;
    estimatedFisherSnedecor = factory.buildAsFisherSnedecor(distribution.getParameter());
    fullprint << "FisherSnedecor from parameters=" << estimatedFisherSnedecor << std::endl;
    // buildMethodOfMoments
    estimatedFisherSnedecor = factory.buildMethodOfMoments(sample);
    fullprint << "Estimated from moments=" << estimatedFisherSnedecor << std::endl;
    const Scalar sample_mu = sample.computeMean()[0];
    const Scalar sample_sigma2 = sample.computeCovariance()(0, 0);
    const Scalar computed_mu = estimatedFisherSnedecor.getMean()[0];
    const Scalar computed_sigma2 = estimatedFisherSnedecor.getCovariance()(0, 0);
    assert_almost_equal(sample_mu, computed_mu, 1e-15, 1e-15);
    assert_almost_equal(sample_sigma2, computed_sigma2, 1e-15, 1e-15);
    // buildMethodOfLikelihoodMaximization
    estimatedFisherSnedecor = factory.buildMethodOfLikelihoodMaximization(sample);
    fullprint << "Estimated from likelihoodMaximization=" << estimatedFisherSnedecor << std::endl;
    const Scalar exact_d1 = distribution.getD1();
    const Scalar exact_d2 = distribution.getD2();
    const Scalar computed_d1 = estimatedFisherSnedecor.getD1();
    const Scalar computed_d2 = estimatedFisherSnedecor.getD2();
    assert_almost_equal(computed_d1, exact_d1, 0.0, 20.0 / sqrt(size));
    assert_almost_equal(computed_d2, exact_d2, 0.0, 20.0 / sqrt(size));
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
